### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by gtrtuugii**

### CODE REVIEW

File: eventbrite.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+            return {
+                "error": error_data.get('error', 'Unknown Error'),
+                "error_description": error_data.get('error_description', 'An error occurred'),
+                "error_detail": error_detail,
+                "status_code": resp.status
+            }
+        else:
+            print(f"HTTP Error: {resp.status}")
+            return None
+    except urllib3.exceptions.HTTPError as e:
```
Comment (robinchew) [inline eventbrite.py]: try block too big, I assume HTTPError is raised by `http.request` so only catch that 1 line:
```py
try:
  resp = http.request(..)
except HttpError:
  pass

if resp.status == 200:
  ...

File: eventbrite.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+import json
+import os
+
+# EVENTBRITE API SCRIPT
+
+# Retrieve the values of an environment variables
+eventbrite_api_token = os.environ.get('EVENTBRITE_API_TOKEN')
+organization_id = os.environ.get('EVENTBRITE_ORGANISATION_ID')
+event_id = os.environ.get('EVENTBRITE_EVENT_ID')
+ticket_class_id = os.environ.get('EVENTBRITE_TICKET_CLASS_ID')
```
Comment (robinchew) [inline eventbrite.py]: Remove event_id, and ticket_class_id, these will be returned by your create event/ticket functions.

File: eventbrite.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+EVENTBRITE_API_BASE_URL = "https://www.eventbriteapi.com/v3/"
+
+# Examples
+event_details = """
```
Comment (robinchew) [inline eventbrite.py]: these should be a dictionaries instead of a string.

File: eventbrite.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+            }
+        else:
+            print(f"HTTP Error: {resp.status}")
+            return None
+    except urllib3.exceptions.HTTPError as e:
+        return {
+            "error": "HTTP Error",
+            "error_description": str(e),
+            "status_code": e.code
+        }
```
Comment (robinchew) [inline eventbrite.py:161]: Create a `quick_create_event` function that calls `createEvent`, `createTicketClass` and `publishEvent`:
```py
def quick_create_event(http, organisation_id, title, description, cost_cents=None, publish=True):
  ticket_detail = {
    "ticket_class": {
      "name": "API Test0",
      "free": true,
      "donation": false,
      "quantity_total": "100",
      "minimum_quantity": "1",
      "maximum_quantity": "10",
      "hidden": false,
      "auto_hide": false,
      "auto_hide_before": "",
      "auto_hide_after": "",
      "sales_channels": ["online","atd"],
      "hide_sale_dates": false,
      "delivery_methods": ["electronic"],
      "include_fee": false,
      "sales_start": "2023-11-09T08:00:00Z",
      "sales_end": "2023-11-11T03:00:00Z",
      **({"cost": f"AUD,{cost_cents}"} if cost is not None else {}),
    }
}
```

File: eventbrite.py (hunk 2 above)
Comment (robinchew) [reply eventbrite.py]: In fact I want all these `os.environ` calls to be run at the bottom where `if __name == '__main__'` is. Pass those values as arguments to your creation functions.
eg.
```py
def create_function(http, api_token, organisation_id, ...)
```

File: eventbrite.py (hunk 4 above)
Comment (robinchew) [reply eventbrite.py:161]: Also need a `date_start` argument. I want to hardcode the venue to be hillview intercultural center https://www.google.com/maps/place/Hillview+Intercultural+Community+Centre/@-32.0090192,115.9003645,17z/data=!3m1!4b1!4m6!3m5!1s0x2a32bdcd5da2c777:0x2fbd272e005ce3d5!8m2!3d-32.0090238!4d115.9029448!16s%2Fg%2F11pckj99n5?entry=ttu

File: eventbrite.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+
+    try:
+        # Make the POST request
+        resp = http.request('POST', url, headers=headers, body=event_details)
+        # Check the status code to ensure a successful request
+        if resp.status == 200:
+            data = resp.data.decode('utf-8')
+            print(data)
+            return json.loads(data)
+        elif resp.status == 400:
```
Comment (robinchew) [inline eventbrite.py]: Remove conditions checking of 200 and 400, assume it's going to be 200. If it's anything but 200 then I want exception to occur.

File: eventbrite.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+
+    # Extract the event_id from the response
+    event_id = new_event.get("id")
+
+    # Create the ticket class details JSON
+    ticket_detail = {
+        "ticket_class": 
+        {
+        "name": "API Test1",
+        "free": "true" if cost_cents is None else "false",
```
Comment (robinchew) [inline eventbrite.py]: put it together with the "cost" key. See below.

File: eventbrite.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+        "auto_hide": "false",
+        "auto_hide_before": "",
+        "auto_hide_after": "",
+        "sales_channels": ["online","atd"],
+        "hide_sale_dates": "false",
+        "delivery_methods": ["electronic"],
+        "include_fee": "false",
+        "sales_start": "2023-11-09T08:00:00Z",
+        "sales_end": "2023-11-11T03:00:00Z",
+        **({"cost": f"AUD,{cost_cents}"} if cost_cents is not None else {}),
```
Comment (robinchew) [inline eventbrite.py]:
```py
        **({"cost": f"AUD,{cost_cents}", "free": "false"} if cost_cents is not None else {'free': "true"}),
```

File: eventbrite.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+    }
+
+    # Convert the event details to a JSON string
+    event_detail_json = json.dumps(event_detail)
+
+    # Create the event
+    new_event = create_event(http, organization_id, event_detail_json)
+
+    # Extract the event_id from the response
+    event_id = new_event.get("id")
```
Comment (robinchew) [inline eventbrite.py]: I'm expecting an actual `event_id` with a non-None value, so don't use get:
```py
event_id = new_event['id']
```

File: eventbrite.py [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+
+    # Create the ticket class details JSON
+    ticket_detail = {
+        "ticket_class":
```
Comment (robinchew) [inline eventbrite.py:403]: Fix your identations.

File: eventbrite.py [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+            "is_series": False,
+            "show_pick_a_seat": False,
+            "show_seatmap_thumbnail": False,
+            "show_colors_in_seatmap_thumbnail": False,
+            "locale": "en_AU"
+        }
+}
+
+paid_ticket_details = {
+        "ticket_class":
```
Comment (robinchew) [inline eventbrite.py]: fix your indentations

File: eventbrite.py [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+
+
+def update_event(http, event_id, event_details):
+    url = f"{EVENTBRITE_API_BASE_URL}events/{event_id}/"
+    headers = {
+    'Authorization': eventbrite_api_token,
+    'Content-Type': 'application/json'
+    }
+
+    try:
```
Comment (robinchew) [inline eventbrite.py]: actually why even catch the error? provided the request sent is correct,  we should expect no errors. If for some reason the usual request don't work anymore, I want this exception to occur and not die silently. Remove the try/except.

File: eventbrite.py [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+            "status_code": resp.status
+            }
+    data = resp.data.decode('utf-8')
+    return json.loads(data)
+
+
+def create_ticket_class(http, event_id, ticket_details):
+    url = f"{EVENTBRITE_API_BASE_URL}events/{event_id}/ticket_classes/"
+    headers = {
+    'Authorization': eventbrite_api_token,
```
Comment (robinchew) [inline eventbrite.py:270]: Don't treat `eventbrite_api_token` as a global, pass it through the function:
```py
def create_ticket_class(http, eventbrite_api_token, event_id, ticket_details)
```
Apply to rest of the functions.

File: eventbrite.py [hunk 13]
```diff
@@ -0,0 +1,422 @@
...
+        return {
+            "error": error_data.get('error', 'Unknown Error'),
+            "error_description": error_data.get('error_description', 'An error occurred'),
+            "error_detail": error_detail,
+            "status_code": resp.status
+        }
+    data = resp.data.decode('utf-8')
+    return json.loads(data)
+
+def quick_create_event(http, eventbrite_api_token, organization_id, title, description, cost_cents=None, publish=True):
```
Comment (robinchew) [inline eventbrite.py]: missing `date_start` (DateTime object) and `duration_hours` (float) arguments. You can use those 2 values to calculate the end datetime.
```py
from datetime import timedelta
date_start + timedelta(hours=duration_hours)
```

File: eventbrite.py [hunk 14]
```diff
@@ -0,0 +1,422 @@
...
+                "utc": "2023-11-30T02:00:00Z"
+            },
+            "currency": "AUD",
+            "online_event": False,
+            "organizer_id": "",
+            "listed": False,
+            "shareable": False,
+            "invite_only": False,
+            "show_remaining": True,
+            "password": "12345",
```
Comment (robinchew) [inline eventbrite.py]: What's this password for?

File: eventbrite.py [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+            },
+            "end": {
+                "timezone": "UTC",
+                "utc": "2023-11-30T02:00:00Z"
+            },
+            "currency": "AUD",
+            "online_event": False,
+            "organizer_id": "",
+            "listed": False,
+            "shareable": False,
```
Comment (robinchew) [inline eventbrite.py]: What makes an event not shareable?

File: eventbrite.py [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+            "maximum_quantity": "10",
+            "hidden": "false",
+            "auto_hide": "false",
+            "auto_hide_before": "",
+            "auto_hide_after": "",
+            "sales_channels": ["online","atd"],
+            "hide_sale_dates": "false",
+            "delivery_methods": ["electronic"],
+            "include_fee": "false",
+            "sales_start": "2023-11-09T08:00:00Z",
```
Comment (robinchew) [inline eventbrite.py]: How does the ticket sales time affect the event start time?

File: eventbrite.py (hunk 15 above)
Comment (gtrtuugii) [reply eventbrite.py]: shareable | boolean | true = Event is shareable, by including social sharing buttons for the Event to Eventbrite applications.

File: eventbrite.py [hunk 17]
```diff
@@ -0,0 +1,422 @@
...
+    # Create the ticket class
+    new_tickets = create_ticket_class(http, eventbrite_api_token, event_id, ticket_detail_json)
+    if publish:
+        publish_event(http,eventbrite_api_token, event_id)
+    return new_event, new_tickets
+
+if __name__ == "__main__":
+
+    # Retrieve the values of an environment variables
+    eventbrite_api_token = os.environ.get('EVENTBRITE_API_TOKEN')
```
Comment (robinchew) [inline eventbrite.py]: these environment variables are mandatory, don't use `.get`

File: eventbrite.py (hunk 14 above)
Comment (gtrtuugii) [reply eventbrite.py]: password | string | Event password used by visitors to access the details of the Event.

File: eventbrite.py [hunk 18]
```diff
@@ -0,0 +1,422 @@
...
+    organization_id = os.environ.get('EVENTBRITE_ORGANIZATION_ID')
+
+    # Create a PoolManager object
+    http = urllib3.PoolManager()
+
+    # quick_create_event(http, eventbrite_api_token, organization_id, "API-TEST-01", "API-TEST-DESCRIPTION", 2000, True)
```
Comment (robinchew) [inline eventbrite.py]: Don't comment this out, perhaps make publish default to False instead. You can even past a mandatory environment variable to control whether to publish or not.

File: eventbrite.py (hunk 16 above)
Comment (gtrtuugii) [reply eventbrite.py]: I believe it's the date when tickets start being available for sale.
sales_start | datetime | Date and time that sales for this Ticket Class begin.

File: eventbrite.py (hunk 18 above)
Comment (robinchew) [reply eventbrite.py]: For the PUBLISH environment variable, that one you can use `.get`. Default can be either true or false.

File: eventbrite.py (hunk 14 above)
Comment (robinchew) [reply eventbrite.py]: where does the visitor enter it?

Comment (robinchew) [reply eventbrite.py]: Make sure the event is made public.

File: eventbrite.py [hunk 19]
```diff
@@ -0,0 +1,430 @@
...
+
+    # Retrieve the values of an environment variables
+    eventbrite_api_token = os.environ['EVENTBRITE_API_TOKEN']
+    organization_id = os.environ['EVENTBRITE_ORGANIZATION_ID']
+    publish = os.environ.get("PUBLISH")
+
+    # Create a PoolManager object
+    http = urllib3.PoolManager()
+
+    quick_create_event(http, eventbrite_api_token, organization_id, "API-TEST-04", "API-TEST-DESCRIPTION-4","2023-11-12T14:00:00Z", 2, 2000, publish)
```
Comment (robinchew) [inline eventbrite.py]: I'm skeptical that the date works if I pass 5pm perth time. eg. "2023-11-12T17:00:00+0800"

Comment (robinchew) [reply eventbrite.py]: I want further additional 'timezone' argument to the `quick_create_event`  function, and I want date_start to be timezone UNAWARE. However the quick_create_event code will combine the time unaware `date_start` and `timezone` argument to be timezone aware datetime object, and then you can convert that to UTC if you want, but it looks like the eventbrite API has a timezone field for dates anyway.

Example timezone in IANA format:  'Australia/Perth'

File: eventbrite.py [hunk 20]
```diff
@@ -87,262 +89,279 @@
...
-            "error": error_data.get('error', 'Unknown Error'),
-            "error_description": error_data.get('error_description', 'An error occurred'),
-            "error_detail": error_detail,
-            "status_code": resp.status
-        }
-    data = resp.data.decode('utf-8')
+    with urllib.request.urlopen(req) as resp:
+        data = resp.read().decode('utf-8')
+
+        if resp.status != 200:
```
Comment (robinchew) [inline eventbrite.py]: Dedent this, once you've read from resp, there's no need to stay within the context manager:
```py
with urllib.request.urlopen():
  data = resp.read().decode()

if resp.status ! = 200:
```

File: eventbrite.py [hunk 21]
```diff
@@ -87,262 +89,279 @@
...
-    # Check the status code to ensure a successful request
+     # Perform the GET request
+    with urllib.request.urlopen(req) as resp:
+        data = resp.read().decode('utf-8')
     if resp.status != 200:
-        print(f"HTTP Error: {resp.status}")
-        
-    data = resp.data.decode('utf-8')
-    print(data)
+        return print(f"HTTP Error: {resp.status}")
```
Comment (robinchew) [inline eventbrite.py:103]: print returns None

**Pull request merged**